<h1>Import Libraries</h1>

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

<h1>Load Dataset into DataFrame</h1>

In [2]:
DataFrame = pd.read_csv('datasets\\uk_road_accident.csv')

In [3]:
DataFrame

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>Check DataFrame Information</h1>

In [4]:
DataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<h1>Basic Descriptive Statistic</h1>

In [5]:
DataFrame.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

<h1>Check and Fill Null Values</h1>

In [6]:
DataFrame.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
DataFrame['Latitude'] = DataFrame['Latitude'].fillna(DataFrame['Latitude'].mode()[0])
DataFrame['Longitude'] = DataFrame['Longitude'].fillna(DataFrame['Longitude'].mode()[0])
DataFrame['Road_Surface_Conditions'] = DataFrame['Road_Surface_Conditions'].fillna(DataFrame['Road_Surface_Conditions'].mode()[0])
DataFrame['Road_Type'] = DataFrame['Road_Type'].fillna(DataFrame['Road_Type'].mode()[0])
DataFrame['Urban_or_Rural_Area'] = DataFrame['Urban_or_Rural_Area'].fillna(DataFrame['Urban_or_Rural_Area'].mode()[0])
DataFrame['Weather_Conditions'] = DataFrame['Weather_Conditions'].fillna(DataFrame['Weather_Conditions'].mode()[0])

In [8]:
DataFrame.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [9]:
DataFrame.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [10]:
DataFrame['Accident Date'] = DataFrame['Accident Date'].astype('category')
#DataFrame['Accident_Severity'] = DataFrame['Accident_Severity'].astype('category')
DataFrame['Light_Conditions'] = DataFrame['Light_Conditions'].astype('category')
DataFrame['District Area'] = DataFrame['District Area'].astype('category')
DataFrame['Road_Surface_Conditions'] = DataFrame['Road_Surface_Conditions'].astype('category')
DataFrame['Road_Type'] = DataFrame['Road_Type'].astype('category')
DataFrame['Urban_or_Rural_Area'] = DataFrame['Urban_or_Rural_Area'].astype('category')
DataFrame['Weather_Conditions'] = DataFrame['Weather_Conditions'].astype('category')


In [11]:
DataFrame.dtypes

Index                        object
Accident_Severity            object
Accident Date              category
Latitude                    float64
Light_Conditions           category
District Area              category
Longitude                   float64
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type                 object
dtype: object

In [12]:
DataFrame['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [13]:
DataFrame['Accident Date'].value_counts()

Accident Date
30-11-2019    704
31-01-2019    697
13-11-2021    692
13-07-2019    692
14-08-2019    688
             ... 
30-12-2022    171
25-12-2019    157
25-12-2022    145
10/1/2022     123
25-12-2020    118
Name: count, Length: 1461, dtype: int64

In [14]:
DataFrame['Latitude'].value_counts()

Latitude
52.458798    75
52.949719    45
53.083165    33
52.967634    31
52.938860    29
             ..
56.880265     1
56.865367     1
57.528413     1
57.194499     1
51.493490     1
Name: count, Length: 511618, dtype: int64

In [15]:
DataFrame['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

In [16]:
DataFrame['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

In [17]:
DataFrame['Longitude'].value_counts()

Longitude
-0.977611    71
-1.871043    36
-0.816789    33
-1.190861    31
-1.216694    29
             ..
-4.169644     1
-3.798648     1
-3.630574     1
-3.836563     1
-3.924477     1
Name: count, Length: 529766, dtype: int64

In [18]:
DataFrame['Number_of_Casualties'].value_counts()

Number_of_Casualties
1     504041
2     107971
3      30776
4      11156
5       4123
6       1525
7        539
8        214
9        112
10        58
11        36
13        24
12        21
14        13
15        11
17         8
19         7
16         7
18         6
22         5
21         4
29         3
40         2
28         2
24         2
27         2
42         2
25         1
32         1
68         1
45         1
62         1
47         1
48         1
26         1
43         1
Name: count, dtype: int64

In [19]:
DataFrame['Number_of_Vehicles'].value_counts()

Number_of_Vehicles
2     391995
1     200787
3      52806
4      11300
5       2464
6        758
7        296
8        147
9         57
10        27
11        12
14         9
12         7
13         6
16         4
28         1
15         1
32         1
19         1
Name: count, dtype: int64

In [20]:
DataFrame['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     448547
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
Name: count, dtype: int64

In [21]:
DataFrame['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

In [22]:
DataFrame['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

In [23]:
DataFrame['Weather_Conditions'].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

In [24]:
DataFrame['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>Clearing any Inconsistencies with the Data Set</h1>

In [25]:
DataFrame['Accident Date'] = DataFrame['Accident Date'].str.strip()
DataFrame['Accident Date'] = DataFrame['Accident Date'].astype('str')
DataFrame['Accident Date'] = DataFrame['Accident Date'].str.replace('/','-')

In [26]:
DataFrame['Accident Date'] = pd.to_datetime(DataFrame['Accident Date'], dayfirst = True, errors = 'coerce')

<h1>Extracting Date Information using Pandas Date Time</h1>

In [27]:
DataFrame['Year'] = DataFrame['Accident Date'].dt.year
DataFrame['Month'] = DataFrame['Accident Date'].dt.month
DataFrame['Day'] = DataFrame['Accident Date'].dt.day
DataFrame['DayOfWeek'] = DataFrame['Accident Date'].dt.dayofweek

In [28]:
DataFrame.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

In [29]:
DataFrame

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  Year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...  

<h1> Explanatory Data Analytics </h1>

<h3> Which District Area Have the Most Serious Accident and How many Fatal Accident happend in that Area?</h3>

In [30]:
DataFrame[DataFrame['Accident_Severity'] == 'Serious']['District Area'].value_counts()

District Area
Birmingham                   1474
Leeds                        1147
Westminster                   926
Bradford                      802
Sheffield                     749
                             ... 
Oswestry                       23
Orkney Islands                 19
Clackmannanshire               18
Shetland Islands               15
London Airport (Heathrow)       9
Name: count, Length: 422, dtype: int64

In [31]:
DataFrame[DataFrame['District Area'] == 'Birmingham']['Accident_Severity'].value_counts()

Accident_Severity
Slight     11912
Serious     1474
Fatal        105
Name: count, dtype: int64

<h3>Insight 1:</h3>
<h5>Birmingham have the most Serious Accident in all Districts and have 105 Fatal Accidents  </h4>
<hr>

<h3> Which Type of Vehicle has the least Slight Accident?</h3>

In [32]:
DataFrame[DataFrame['Accident_Severity'] == 'Slight']['Vehicle_Type'].min()

'Agricultural vehicle'

<h3>Insight 2:</h3>
<h5> The Agricultral Vehicles have the least of Slight Accident. </h4>
<hr>

<h3>. Do accidents with more vehicles usually have more casualties?  </h3>

In [33]:
DataFrame[['Number_of_Vehicles','Number_of_Casualties']].corr()

Number_of_Vehicles  Number_of_Casualties
Number_of_Vehicles              1.000000              0.228889
Number_of_Casualties            0.228889              1.000000

<h3>Insight 3:</h3>
<h5> A positive correlation confirms that multi-vehicle accidents increase casualty counts.</h4>
<hr>

<h3> Are latitude and longitude correlated?</h3>

In [34]:
DataFrame[['Latitude','Longitude']].corr()

Latitude  Longitude
Latitude   1.000000  -0.398114
Longitude -0.398114   1.000000

<h3>Insight 4:</h3>
<h5>They are usually negatively correlated because of how geographic coordinates are distributed.</h4>
<hr>

<h3>Which day of the week has the highest average casualties?</h3> 

In [35]:
DataFrame.groupby('DayOfWeek')['Number_of_Casualties'].mean()

DayOfWeek
0    1.465685
1    1.331793
2    1.317925
3    1.321160
4    1.323238
5    1.350977
6    1.423272
Name: Number_of_Casualties, dtype: float64

<h3>Insight 5:</h3>
<h5>Weekends may show higher averages due to faster driving and leisure trips.</h4>
<hr>

<h3> Which road type records the most accidents?</h3> 

In [36]:
DataFrame['Road_Type'].value_counts()

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h3>Insight 6:</h3>
<h5>Single carriageway roads show the highest accident counts, likely due to high usage.</h4>
<hr>

<h3>Do urban or rural accidents have more casualties on average?</h3> 

In [37]:
DataFrame.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287807
Name: Number_of_Casualties, dtype: float64

<h3>Insight 7:</h3>
<h5>Rural accidents have slightly higher casualties.</h4>
<hr>

<h3>Do accidents in daylight or darkness involve more vehicles on average?</h3> 

In [38]:
DataFrame.groupby('Light_Conditions')['Number_of_Vehicles'].mean()

Light_Conditions
Darkness - lighting unknown    1.758482
Darkness - lights lit          1.775498
Darkness - lights unlit        1.752654
Darkness - no lighting         1.599781
Daylight                       1.865385
Name: Number_of_Vehicles, dtype: float64

<h3>Insight 8:</h3>
<h5>Accidents at night may involve fewer vehicles but can be more severe due to visibility issues.</h4>
<hr>

<h3>Which year had the highest average casualties per accident?</h3> 

In [39]:
DataFrame.groupby('Year')['Number_of_Casualties'].mean()

Year
2019    1.360569
2020    1.353559
2021    1.358243
2022    1.355341
Name: Number_of_Casualties, dtype: float64

<h3>Insight 9:</h3>
<h5>2019 had the highest average but its not too far on the few years.</h4>
<hr>

<h3>How many non-missing records exist in Light_Conditions?</h3> 

In [40]:
DataFrame['Light_Conditions']

0               Darkness - lights lit
1                            Daylight
2         Darkness - lighting unknown
3                            Daylight
4               Darkness - lights lit
                     ...             
660674                       Daylight
660675         Darkness - no lighting
660676                       Daylight
660677         Darkness - no lighting
660678                       Daylight
Name: Light_Conditions, Length: 660679, dtype: category
Categories (5, object): ['Darkness - lighting unknown', 'Darkness - lights lit', 'Darkness - lights unlit', 'Darkness - no lighting', 'Daylight']

In [41]:
DataFrame['Light_Conditions'].count()

np.int64(660679)

<h3>Insight 10:</h3>
<h5>Indicates data completeness.</h4>
<hr>

<h3>Yearly Accidents that happen in London</h3> 

In [42]:
DataFrame[DataFrame['District Area'] == 'City of London']['Year'].describe()

count    1317.000000
mean     2020.477601
std         1.130063
min      2019.000000
25%      2019.000000
50%      2020.000000
75%      2022.000000
max      2022.000000
Name: Year, dtype: float64

<h3>Insight 11:</h3>
<h5>2019 is the minimum of accidents in london while the maximum at 2022.</h4>
<hr>

<h3>Insight 12:</h3>
<h5>The average accident that happen in London is in the Year of 2020.</h4>
<hr>

<h3>Insight 13:</h3>
<h5>The total Accidents that happen in London is 1317.</h4>
<hr>

<h3>Monthly Accidents that happen in the Roundabout Roads</h3> 

In [43]:
DataFrame[DataFrame['Vehicle_Type'] == 'Car']['Road_Type'].value_counts()

Road_Type
Single carriageway    374464
Dual carriageway       74820
Roundabout             33187
One way street         10223
Slip road               5298
Name: count, dtype: int64

<h3>Insight 14:</h3>
<h5>Mostly Car are prone to accident in Singgle Carriageway</h4>
<hr>

<h3>Insight 15:</h3>
<h5>Least Car having accident in Slip road maybe because they know its slippery road its need more careful to drive</h4>
<hr>

<h3>Insight 16:</h3>
<h5>Its not about the road who will tell the highest possibility of accident</h4>
<hr>

<h3>Average accident that happen on Kensington and Chelsea in every month of 2020</h3>

In [44]:
kensington_2020 = DataFrame[(DataFrame['District Area'] == 'Kensington and Chelsea') & (DataFrame['Year'] == 2020)]['Month'].mean()

In [45]:
kensington_2020

np.float64(6.602739726027397)

<h3>Insight 17:</h3>
<h5>The Average Accident that happened in Kensington and Chelsea is 6.602739726027397.</h4>
<hr>

<h3>Average Casualties per month</h3> 

In [46]:
DataFrame.groupby('Month')['Number_of_Casualties'].mean()

Month
1     1.343585
2     1.343941
3     1.351107
4     1.364796
5     1.362791
6     1.350826
7     1.366350
8     1.396324
9     1.347144
10    1.347801
11    1.348024
12    1.362972
Name: Number_of_Casualties, dtype: float64

<h3>Insight 18:</h3>
<h5>August was the Average Casualties per month</h4>
<hr>

In [47]:
DataFrame

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  Year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...  

<h3>What is Most Casualties in every Types of Vehicles and the average of it?</h3> 

In [48]:
DataFrame.groupby('Vehicle_Type') ['Number_of_Casualties'].max()

Vehicle_Type
Agricultural vehicle                      8
Bus or coach (17 or more pass seats)     62
Car                                      68
Data missing or out of range              2
Goods 7.5 tonnes mgw and over            19
Goods over 3.5t. and under 7.5t          21
Minibus (8 - 16 passenger seats)          9
Motorcycle 125cc and under               13
Motorcycle 50cc and under                12
Motorcycle over 125cc and up to 500cc    11
Motorcycle over 500cc                    27
Other vehicle                             9
Pedal cycle                               5
Ridden horse                              2
Taxi/Private hire car                    25
Van / Goods 3.5 tonnes mgw or under      48
Name: Number_of_Casualties, dtype: int64

In [49]:
DataFrame.groupby('Vehicle_Type') ['Number_of_Casualties'].max().mean()

np.float64(21.3125)

<h3>Insight 19:</h3>
<h5>Looks like a highest Vehicle type that have casualties was also a Car not only in generally</h4>
<hr>

<h3>Insight 20:</h3>
<h5>The average of that casualties is 21</h4>
<hr>

<h3>Serious that happened on the District Areas</h3> 

In [50]:
DataFrame[DataFrame['Accident_Severity']== 'Serious'].groupby('District Area').size()

District Area
Aberdeen City    239
Aberdeenshire    463
Adur             101
Allerdale        143
Alnwick           33
                ... 
Wychavon         193
Wycombe          216
Wyre             186
Wyre Forest      132
York             255
Length: 422, dtype: int64

In [51]:
DataFrame[DataFrame['Accident_Severity']== 'Serious'].groupby('District Area').size().max()

np.int64(1474)

In [52]:
DataFrame[DataFrame['Accident_Severity']== 'Serious'].groupby('District Area').size().min()

np.int64(9)

In [53]:
DataFrame[DataFrame['Accident_Severity']== 'Serious'].groupby('District Area').size().mean()

np.float64(209.04502369668248)

<h3>Insight 21:</h3>
<h5>The Highest Accident occured in District areas in 1474</h4>
<hr>

<h3>Insight 22:</h3>
<h5>The Least Accident occured in District areas in 9</h4>
<hr>

<h3>Insight 23:</h3>
<h5>The average happend on that Disctrict areas is 209.04502369668248.</h4>
<hr>

In [54]:
DataFrame[DataFrame['Accident_Severity']== 'Fatal'].groupby('Road_Type').size()

Road_Type
Dual carriageway      1815
One way street          95
Roundabout             142
Single carriageway    6560
Slip road               49
dtype: int64

In [55]:
DataFrame[DataFrame['Accident_Severity']== 'Fatal'].groupby('Road_Type').size().max()

np.int64(6560)

In [56]:
DataFrame[DataFrame['Accident_Severity']== 'Fatal'].groupby('Road_Type').size().min()

np.int64(49)

<h3>Insight 24:</h3>
<h5>Highest Fatal accident occured in different Road Type is Single carriage with a total of 6560.</h4>
<hr>

<h3>Insight 25:</h3>
<h5>Highest Fatal accident occured in different Road Type is Sliproad with a total of 49..</h4>
<hr>